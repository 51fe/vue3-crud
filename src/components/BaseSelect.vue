<script setup lang="ts">
import { computed } from 'vue'

export interface BaseSelectProps {
  modelValue?: SelectValue
  options: Option[]
  multiple?: boolean
  placeholder?: string
}

const props = withDefaults(
  defineProps<BaseSelectProps>(),
  {
    modelValue: undefined,
    placeholder: '请选择'
  }
)

const emit = defineEmits<{
  (e: 'update:modelValue', value: SelectValue): void
}>()

const toNumber = (str: string) => isNaN?.(Number(str)) ? str : Number(str)

const current = computed({
  get: () => {
    const value = props.modelValue
    if (value && typeof value === 'string') {
      if (props.multiple) {
        return value.split(',').map(item => toNumber(item))
      }
      return toNumber(value)
    }
    return value
  },
  set: (value) => {
    if(props.multiple && Array.isArray(value)) {
      value = value.join(',')
    }
    emit('update:modelValue', value as SelectValue)
  }
})
</script>

<template>
  <el-select
    v-model="current"
    :multiple="multiple"
    :placeholder="placeholder"
    filterable
    clearable
    class="base-select"
  >
    <el-option
      v-for="item in options"
      :key="item.value"
      :label="item.label"
      :value="item.value"
    />
  </el-select>
</template>

<style>
.base-select {
  width: 100%;
}
</style>
