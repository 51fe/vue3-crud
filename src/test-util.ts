import { render, RenderOptions, RenderResult } from '@testing-library/vue'

import ElementPlus from 'element-plus'
import CN from 'element-plus/lib/locale/lang/zh-cn'

export const customRender = (
  ui: any, overrides?: Omit<RenderOptions, 'global'>): RenderResult => {
  return render(ui, {
    global: {
      plugins: [[ElementPlus, ...[{ locale:  CN }]]]
    },
    ...overrides
  })
}
export { defineComponent, ref, reactive, nextTick } from 'vue'
export * from '@testing-library/vue'
export { customRender as render }