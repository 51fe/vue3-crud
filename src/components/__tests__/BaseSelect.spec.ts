import { fireEvent, screen, render } from '../../test-util'

import BaseSelect from '../BaseSelect.vue'

const options = [{
  value: 1,
  label: '移动'
}, {
  value: 2,
  label: '联通'
}, {
  value: 3,
  label: '电信'
}]

const defaultPlaceholder = '请选择'

const setUp = (overrides?: any) => {
  return render(BaseSelect, {
    props: {
      options,
      ...overrides
    }
  })
}

const getOptionByLabel = (label: string) => screen.getByRole('listitem', {
  hidden: true,
  name: (content, el) => el.textContent === label
})

test('renders correctly by default', async () => {
  const { html } = setUp()
  screen.getByPlaceholderText(defaultPlaceholder)
  options.forEach((item) => {
    getOptionByLabel(item.label)
  })
  const list = screen.getAllByRole('listitem', { hidden: true })
  expect(list).toHaveLength(options.length)
  expect(html()).toMatchSnapshot()
})

test('changes props correctly', async () => {
  const placeholder = '请选择营运商'
  const modelValue = 3
  setUp({
    placeholder,
    modelValue
  })
  // change placeholder and current selected
  await screen.findByPlaceholderText(placeholder)
  expect(getOptionByLabel('电信')).toHaveClass('selected')
})

test('emits input event correctly when selected', async () => {
  const { emitted } = setUp()
  //  trigger an update event by clicking the <option> element.
  await fireEvent.click(getOptionByLabel('电信'))
  expect(emitted()['update:modelValue'][0]).toEqual([3])
})

test('Passes and returns a string when select multiple items', async () => {
  const modelValue = '1, 2'
  const { emitted } = setUp({
    multiple: true,
    modelValue
  })
  // current selected 移动, 联通
  modelValue.split(',').forEach(id => {
    const found = options.find(item => item.value === parseInt(id))
    if (found) {
      getOptionByLabel(found.label)
    }
  })
  // select all
  await fireEvent.click(getOptionByLabel('电信'))
  expect(emitted()['update:modelValue'][0]).toEqual(['1,2,3'])
})




