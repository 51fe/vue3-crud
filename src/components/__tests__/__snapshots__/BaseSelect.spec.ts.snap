// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly by default 1`] = `
"<div class="el-select base-select">
  <div class="select-trigger el-tooltip__trigger el-tooltip__trigger">
    <!--v-if-->
    <div class="el-input el-input--suffix">
      <!-- input -->
      <!-- prepend slot -->
      <!--v-if-->
      <div class="el-input__wrapper">
        <!-- prefix slot -->
        <!--v-if--><input class="el-input__inner" type="text" readonly="" autocomplete="off" tabindex="0" placeholder="请选择"><!-- suffix slot --><span class="el-input__suffix"><span class="el-input__suffix-inner"><i class="el-icon el-select__caret el-select__icon"><svg viewBox="0 0 1024 1024" xmlns="http://www.w3.org/2000/svg"><path fill="currentColor" d="M831.872 340.864 512 652.672 192.128 340.864a30.592 30.592 0 0 0-42.752 0 29.12 29.12 0 0 0 0 41.6L489.664 714.24a32 32 0 0 0 44.672 0l340.288-331.712a29.12 29.12 0 0 0 0-41.728 30.592 30.592 0 0 0-42.752 0z"></path></svg></i><!--v-if--><!--v-if--><!--v-if--><!--v-if--><!--v-if--><!--v-if--></span></span>
      </div><!-- append slot -->
      <!--v-if-->
    </div>
  </div>
  <!--teleport start-->
  <!--teleport end-->
</div>"
`;
